File: Bachelor-Thesis/legacy/cosserat_rod/cosserat_rod.py
import numpy as np

from utils.tube import Tube
from utils.optimization import Function
from solvers.solvers import Solver
from integrators.integrators import Integrator


class CosseratRod(Function):
    """
    Cosserat Rod Modelling Algorithm for a 3-segment CTCR
    """
    n = 3

    def __init__(self, ctcr: list[Tube], joints: np.ndarray, solver: Solver, integrator: Integrator) -> None:
        if len(ctcr) != self.n:
            raise Exception(f"CTCR must have {self.n} segments.")

        self.idx = 0
        self._shooting_iter = 0  # Counter for how many times derive has been called

        self.y = np.zeros((0, ))

        self.KBT = np.array([c.kbt for c in ctcr])
        self.U = np.array([c.u for c in ctcr])

        self.ALPHAS = joints[:, 0]
        self.BETAS = joints[:, 1]

        self.L = np.array([c.L for c in ctcr])
        self.Ls = np.array([c.Ls for c in ctcr])

        self.ctcr = ctcr

        self._initialize()

        self.state_idx = 12

        self.solver = solver
        self.integrator = integrator

        self.backbone = []


    def _initialize(self):
        self.tube_ends = [self.ctcr[i].L + self.BETAS[i] for i in range(self.n)]

        self.tube_lengths = [0,] + self.tube_ends +\
            [self.ctcr[i].Ls + self.BETAS[i] for i in range(self.n)]
        self.tube_lengths = sorted(self.tube_lengths)

        self.idx = self.tube_lengths.index(0)


    def _construct_p_rot(self, psi_0: float) -> np.ndarray:
        c_psi = np.cos(psi_0)
        s_psi = np.sin(psi_0)

        p_init = np.zeros((3, ))

        R_init = np.array([
            [c_psi, -s_psi, 0],
            [s_psi,  c_psi, 0],
            [    0,      0, 1]
        ]).reshape((9,))

        return np.concatenate((p_init, R_init))


    def _construct_p_rot_dot(self, y: np.ndarray, u0: np.ndarray) -> np.ndarray:
        R = np.reshape(y[3:12], (3, 3))

        p = R[:, 2]

        u_hat = np.array([
            [     0, -u0[2],  u0[1]],
            [ u0[2],      0, -u0[0]],
            [-u0[1],  u0[0],      0]
        ])
        R_dot = R@u_hat

        return np.concatenate((p, R_dot.reshape((9,))))


    def _is_tube(self, i: int) -> bool:
        return (self.tube_lengths[self.idx] + self.tube_lengths[self.idx + 1])/2 < self.BETAS[i] + self.L[i]


    def _is_straight(self, i: int) -> bool:
        return (self.tube_lengths[self.idx] + self.tube_lengths[self.idx + 1])/2 < self.BETAS[i] + self.Ls[i]


    def construct_tip(self) -> np.ndarray:
        tip = np.zeros((4, 4))

        # Construct Rotation (Orientation)
        tip[:3, :3] = self.y[3:12].reshape((3, 3))

        # Construct Position
        tip[:3, 3] = self.y[:3]
        tip[3, 3] = 1

        return tip


    def derive(self, t: float, y: np.ndarray) -> np.ndarray:
        # Based on: https://ieeexplore.ieee.org/document/5559519
        # https://ieeexplore.ieee.org/document/5980351

        self._shooting_iter += 1

        uz = y[self.state_idx:self.state_idx + self.n]
        psi = y[self.state_idx + self.n:self.state_idx + 2*self.n]

        theta = psi[1:] - psi[0]
        c_theta = np.cos(theta)
        s_theta = np.sin(theta)

        Rtheta = np.zeros((self.n - 1, 3, 3))
        Rtheta[:, 0, 0] = c_theta
        Rtheta[:, 0, 1] = -s_theta
        Rtheta[:, 1, 0] = s_theta
        Rtheta[:, 1, 1] = c_theta
        Rtheta[:, 2, 2] = 1

        e3 = np.array([0, 0, 1])
        u = np.zeros((self.n, 3))

        # Get tube info
        u_star = np.zeros((self.n, 3))

        if not self._is_straight(0):
            u_star[0, :] = self.ctcr[0].u

        k = np.zeros((3, 3))

        ks = np.zeros((self.n, 3, 3))
        ks[0, :, :] = self.ctcr[0].kbt
        k += ks[0]

        u[0] = ks[0]@u_star[0]
        for i in range(1, self.n):
            if self._is_tube(i) and not self._is_straight(i):
                u_star[i, :] = self.ctcr[i].u

            if self._is_tube(i):
                ks[i, :, :] = self.ctcr[i].kbt

            k += ks[i]

            u[0] += Rtheta[i - 1]@ks[i]@u_star[i] - ks[i][2, 2]*(uz[i] - uz[0])*e3

        invk = np.diag(1/np.diag(k))
        u[0] = invk@u[0]

        uz_dot = np.zeros((self.n,))
        uz_dot[0] = ks[0][0, 0]/ks[0][2, 2]*(u[0][0]*u_star[0][1] - u[0][1]*u_star[0][0])

        for i in range(1, self.n):
            u[i] = Rtheta[i - 1].T@u[0] + (uz[i] - uz[0])*e3

            if (self.tube_lengths[self.idx] + self.tube_lengths[self.idx + 1])/2 < self.BETAS[i] + self.ctcr[i].L:
                uz_dot[i] = ks[i][0, 0]/ks[i][2, 2]*(u[i][0]*u_star[i][1] - u[i][1]*u_star[i][0])

        p_rot_dot = self._construct_p_rot_dot(y, u[0])

        return np.concatenate((p_rot_dot, uz_dot, uz, (1, )))


    def evaluate(self, x: np.ndarray) -> np.ndarray:
        # Based on: https://ieeexplore.ieee.org/document/5980351
        self.backbone = [[]]

        psi_init = self.ALPHAS + x*np.abs(self.BETAS)
        p_rot_init = self._construct_p_rot(psi_init[0])

        self.y = np.concatenate((p_rot_init, x, psi_init, (0,)))

        start_idx = self.idx
        while self.idx < 2*self.n:
            if self.tube_lengths[self.idx] in self.tube_ends:
                self.backbone.append([])

            self.y = self.integrator.integrate(self, self.y, self.tube_lengths[self.idx], self.tube_lengths[self.idx + 1])
            self.backbone[-1].extend(np.array(self.integrator.ys)[:, :self.state_idx].tolist())  # Add the position and orientation of backbone

            self.idx += 1
        self.idx = start_idx

        return self.y[self.state_idx:self.state_idx + self.n]


    def solve(self, u_init_guess: np.ndarray) -> np.ndarray:
        self._shooting_iter = 0
        return self.solver.solve(self, u_init_guess)

File: Bachelor-Thesis/src/CosseratRod.cpp
#include "../include/CosseratRod.hpp"

using namespace std::placeholders;


CosseratRod::CosseratRod(Tube ctcr[TUBE_NUM]){
    this->init(ctcr);
}


CosseratRod::CosseratRod(Tube ctcr[TUBE_NUM], Eigen::Matrix<DTYPE_t, TUBE_NUM, 2> joints, u_char solver, u_char integrator, DTYPE_t h, bool quaternion, DTYPE_t fd_step){
    this->init(ctcr);

    this->quaternion_mode = quaternion;

    this->solver.setSolver(solver);
    this->integrator.setIntegrator(integrator);

    this->h = h;
    this->fd_step = fd_step;

    this->setJoints(joints);

    this->Bb.setIdentity();
    this->Bb.diagonal().head(TUBE_NUM - 1).setZero();
}


void CosseratRod::init(Tube ctcr[TUBE_NUM]){
    this->ctcr = ctcr;

    this->_derive = std::bind(&CosseratRod::derive, this, _1, _2);
    this->_evaluate = std::bind(&CosseratRod::evaluate, this, _1, _2);
}


Eigen::Matrix<DTYPE_t, TUBE_NUM, 2> CosseratRod::getJoints(){
    Eigen::Matrix<DTYPE_t, 3, 2> joints;
    joints.col(0) = this->alphas;
    joints.col(1) = this->betas;

    return joints;
}


void CosseratRod::setJoints(Eigen::Matrix<DTYPE_t, TUBE_NUM, 2> joints){
    this->alphas = joints.col(0);
    this->betas = joints.col(1);

    this->tube_lengths[0] = 0;
    for(int i = 1; i < TUBE_NUM + 1; i++){
        this->tube_ends[i - 1] = ctcr[i - 1].l + this->betas(i - 1);

        this->tube_lengths[i] = ctcr[i - 1].ls + this->betas(i - 1);
        this->tube_lengths[TUBE_NUM + i] = ctcr[i - 1].l + this->betas(i - 1);
    }

    std::sort(this->tube_lengths, this->tube_lengths + 2*TUBE_NUM + 1);

    for(int i = 0; i < 2*TUBE_NUM + 1; i++){
        if(this->tube_lengths[i] == 0){
            this->idx = i;
            break;
        }
    }
}


Eigen::Vector<DTYPE_t, P_ROT_NUM_Q> CosseratRod::constructPRotInitQ(DTYPE_t psi_0){
    // Quaternion Model
    Eigen::Vector<DTYPE_t, 4> q_init;
    q_init << (DTYPE_t)cos(psi_0/2.0),
              0,
              0,
              (DTYPE_t)sin(psi_0/2.0);

    Eigen::Vector<DTYPE_t, P_ROT_NUM_Q> p_rot;
    p_rot << Eigen::Vector<DTYPE_t, 3>::Zero(),
             q_init;

    return p_rot;
}


Eigen::Vector<DTYPE_t, P_ROT_NUM_Q> CosseratRod::constructPRotDotQ(VectorState &y0, VectorSpace &u0){
    // Quaternion Model
    Eigen::Matrix<DTYPE_t, 4, 3> Wt;
    Wt << -y0(4), -y0(5), -y0(6),
           y0(3), -y0(6),  y0(5),
           y0(6),  y0(3), -y0(4),
          -y0(5),  y0(4),  y0(3);

    Eigen::Vector<DTYPE_t, P_ROT_NUM_Q> p_rot_dot;
    p_rot_dot << 2*(y0(4)*y0(6) + y0(3)*y0(5)),
             2*(y0(5)*y0(6) - y0(3)*y0(4)),
             y0(3)*y0(3) - y0(4)*y0(4) - y0(5)*y0(5) + y0(6)*y0(6),
             Wt*u0/2;  // q_dot = W.T*u0/2

    return p_rot_dot;
}


Eigen::Vector<DTYPE_t, P_ROT_NUM_R> CosseratRod::constructPRotInitR(DTYPE_t psi_0){
    DTYPE_t c_psi = cos(psi_0);
    DTYPE_t s_psi = sin(psi_0);

    Eigen::Matrix<DTYPE_t, 3, 3> R_init;
    R_init << c_psi, -s_psi, 0,
              s_psi,  c_psi, 0,
                  0,      0, 1;

    Eigen::Vector<DTYPE_t, P_ROT_NUM_R> p_rot;
    p_rot << Eigen::Vector<DTYPE_t, 3>::Zero(),
             R_init.reshaped(9, 1);

    return p_rot;
}


Eigen::Vector<DTYPE_t, P_ROT_NUM_R> CosseratRod::constructPRotDotR(VectorState &y0, VectorSpace &u0){
    Eigen::Matrix<DTYPE_t, 3, 3> R = y0.segment(3, 9).reshaped(3, 3);

    Eigen::Matrix<DTYPE_t, 3, 3> R_dot;
    R_dot <<     0, -u0(2),  u0(1),
             u0(2),      0, -u0(0),
            -u0(1),  u0(0),      0;  // u_hat
    R_dot = R*R_dot;

    Eigen::Vector<DTYPE_t, P_ROT_NUM_R> p_rot_dot;
    p_rot_dot << R.col(2),
             R_dot.reshaped(9, 1);

    return p_rot_dot;
}


Eigen::Matrix<DTYPE_t, 4, 4> CosseratRod::constructTubeEnd(uint tube_end){
    Eigen::Matrix<DTYPE_t, 4, 4> tip;
    tip.setZero();

    VectorState s = this->tube_states[tube_end];

    if(this->quaternion_mode){ // Quaternion Model
        tip(0, 0) = s(3)*s(3) + s(4)*s(4) - s(5)*s(5) - s(6)*s(6);
        tip(1, 0) = 2 * (s(4)*s(5) + s(3)*s(6));
        tip(2, 0) = 2 * (s(4)*s(6) - s(3)*s(5));

        tip(0, 1) = 2 * (s(4)*s(5) - s(3)*s(6));
        tip(1, 1) = s(3)*s(3) - s(4)*s(4) + s(5)*s(5) - s(6)*s(6);
        tip(2, 1) = 2 * (s(5)*s(6) - s(3)*s(4));

        tip(0, 2) = 2 * (s(4)*s(6) + s(3)*s(5));
        tip(1, 2) = 2 * (s(5)*s(6) + s(3)*s(4));
        tip(2, 2) = s(3)*s(3) - s(4)*s(4) - s(5)*s(5) + s(6)*s(6);

        // Construct Position
        tip.col(3) << s.head(3),
                    1;
    }
    else{  // Rotation Model
        tip.block<3, 3>(0, 0) = s.segment(3, 9).reshaped(3, 3);
        tip.col(3) << s.head(3),
                    1;
    }

    return tip;
}


bool CosseratRod::isTube(int i){
    return (this->tube_lengths[this->idx] + this->tube_lengths[this->idx + 1])/2 < this->betas(i) + this->ctcr[i].l;
}


bool CosseratRod::isStraight(int i){
    return (this->tube_lengths[this->idx] + this->tube_lengths[this->idx + 1])/2 < this->betas(i) + this->ctcr[i].ls;
}


StateVariables CosseratRod::getStateVariables(VectorState &y0){
    StateVariables state;

    // Prepare info to be used
    VectorInput psi;

    if(this->quaternion_mode){
        state.uz = y0.segment(P_ROT_NUM_Q, 3);
        psi = y0.segment(P_ROT_NUM_Q + 3, 3);
    }else{
        state.uz = y0.segment(P_ROT_NUM_R, 3);
        psi = y0.segment(P_ROT_NUM_R + 3, 3);
    }

    // Prepare Tube Info
    if(!this->isStraight(0)){
        state.u_star[0] = this->ctcr[0].u;
    }
    else{
        state.u_star[0].setZero();
    }

    state.ks[0] = this->ctcr[0].kbt;
    state.u0 = state.ks[0]*state.u_star[0];

    state.theta[0] = 0;
    state.K = state.ks[0];

    DTYPE_t c_theta, s_theta;
    for(int i = 1; i < TUBE_NUM; i++){
        state.theta[i] = psi(i) - psi(0);

        c_theta = cos(state.theta[i]);
        s_theta = sin(state.theta[i]);

        state.Rtheta[i - 1] << c_theta, -s_theta, 0,
                         s_theta,  c_theta, 0,
                         0,        0,       1;

        if(this->isTube(i) && !this->isStraight(i)){
            state.u_star[i] = this->ctcr[i].u;
        }
        else{
            state.u_star[i].setZero();
        }
        
        if(this->isTube(i)){
            state.ks[i] = this->ctcr[i].kbt;
        }
        else{
            state.ks[i].setZero();
        }

        state.K += state.ks[i];
        state.u0 += state.Rtheta[i - 1]*state.ks[i]*state.u_star[i] - state.ks[i](2, 2)*(state.uz[i] - state.uz[0])*e3;
    }

    state.u0 = state.u0.cwiseQuotient(state.K.diagonal());

    return state;
}


void CosseratRod::constructJacobian(VectorState &y0){
    this->A.setZero();
    this->A.col(TUBE_NUM + 1).head(TUBE_NUM - 1).setConstant(-1);
    this->A.topRightCorner(TUBE_NUM - 1, TUBE_NUM - 1).diagonal().setConstant(1);

    StateVariables state = this->getStateVariables(y0);

    DTYPE_t c_theta, s_theta;
    for(int i = 0; i < TUBE_NUM; i++){
        c_theta = cos(state.theta[i]);
        s_theta = sin(state.theta[i]);

        if(i > 0){
            // d u_dot_1,x / d theta_i
            this->A(TUBE_NUM - 1, i - 1) = state.uz[i]*state.u0[0]*c_theta + state.uz[i]*state.u0[1]*s_theta;
            this->A(TUBE_NUM - 1, i - 1) *= state.ks[i](0, 0);

            this->A(TUBE_NUM - 1, i - 1) += state.ks[i](2, 2)*(state.uz[i] - state.u_star[i][2])*(-state.u0[0]*c_theta - state.u0[1]*s_theta);

            // d u_dot_1,y / d theta_i
            this->A(TUBE_NUM, i - 1) = -state.uz[i]*state.u0[0]*s_theta + state.uz[i]*state.u0[1]*c_theta;
            this->A(TUBE_NUM, i - 1) *= state.ks[i](0, 0);

            this->A(TUBE_NUM, i - 1) -= state.ks[i](2, 2)*(state.uz[i] - state.u_star[i][2])*(-state.u0[0]*s_theta + state.u0[1]*c_theta);
        }

        // d u_dot_1,x / d u_1,x and d u_dot_1,x / d u_1,y
        this->A(TUBE_NUM - 1, TUBE_NUM - 1) += state.ks[i](0, 0)*(state.uz[i]*s_theta) - state.ks[i](2, 2)*s_theta*(state.uz[i] - state.u_star[i][2]);
        this->A(TUBE_NUM - 1, TUBE_NUM) += state.ks[i](0, 0)*(state.uz[i] - state.uz[0] - state.uz[i]*c_theta) + state.ks[i](2, 2)*c_theta*(state.uz[i] - state.u_star[i][2]);

        // d u_dot_1,x / d u_i,z
        this->A(TUBE_NUM - 1, TUBE_NUM + 1 + i) = state.u0[1] + state.u0[0]*s_theta - state.u0[1]*c_theta + state.u_star[i][1];
        this->A(TUBE_NUM - 1, TUBE_NUM + 1 + i) *= state.ks[i](0, 0);

        this->A(TUBE_NUM - 1, TUBE_NUM + 1 + i) -= state.ks[i](2, 2)*(state.u0[0]*s_theta - state.u0[1]*c_theta);

        // d u_dot_1,y / d u_i,z
        this->A(TUBE_NUM, TUBE_NUM + 1 + i) = -state.u0[0] + state.u0[0]*c_theta + state.u0[1]*s_theta - state.u_star[i][0];
        this->A(TUBE_NUM, TUBE_NUM + 1 + i) *= state.ks[i](0, 0);

        this->A(TUBE_NUM, TUBE_NUM + 1 + i) -= state.ks[i](2, 2)*(state.u0[0]*c_theta + state.u0[1]*s_theta);

        // d u_dot_i,z / d theta_i
        if(i > 0){
            this->A(TUBE_NUM + 1 + i, i - 1) = state.u_star[i][1]*(-state.u0[0]*s_theta + state.u0[1]*c_theta);
            this->A(TUBE_NUM + 1 + i, i - 1) += state.u_star[i][0]*(state.u0[0]*c_theta + state.u0[1]*s_theta);
        }

        // d u_dot_i,z / d u_1,x and d u_dot_i,z / d u_1,y
        this->A(TUBE_NUM + 1 + i, TUBE_NUM - 1) = state.u_star[i][1]*c_theta + state.u_star[i][0]*s_theta;
        this->A(TUBE_NUM + 1 + i, TUBE_NUM) = state.u_star[i][1]*s_theta - state.u_star[i][0]*c_theta;

        // Multiply by EI/GJ
        this->A.row(TUBE_NUM + 1 + i) *= state.ks[i](0, 0)/state.ks[i](2, 2);
    }

    this->A.row(TUBE_NUM - 1) /= -state.K(0, 0);
    this->A.row(TUBE_NUM) /= -state.K(0, 0);

    this->A(TUBE_NUM, TUBE_NUM) = this->A(TUBE_NUM - 1, TUBE_NUM - 1);
    this->A(TUBE_NUM, TUBE_NUM - 1) = -this->A(TUBE_NUM - 1, TUBE_NUM);
}


DTYPE_t CosseratRod::calculateMaxJacobianEigen(VectorState y0){
    this->constructJacobian(y0);
    return this->A.eigenvalues().cwiseAbs().maxCoeff();
}


VectorState CosseratRod::derive(DTYPE_t t, VectorState &y0){
    if(this->calculate_J){
        this->J_eigen = std::max(this->J_eigen, this->calculateMaxJacobianEigen(y0));
    }

    this->func_counter++;

    VectorInput uz_dot;
    VectorSpace u[TUBE_NUM - 1];
    StateVariables state = this->getStateVariables(y0);

    uz_dot(0) = state.ks[0](0, 0)/state.ks[0](2, 2)*(state.u0(0)*state.u_star[0](1) - state.u0(1)*state.u_star[0](0));
    for(int i = 1; i < TUBE_NUM; i++){
        u[i] = state.Rtheta[i - 1].transpose()*state.u0 + (state.uz(i) - state.uz(0))*e3;

        if(this->isTube(i)){
            uz_dot(i) = state.ks[i](0, 0)/state.ks[i](2, 2)*(u[i](0)*state.u_star[i](1) - u[i](1)*state.u_star[i](0));
        }
        else{
            uz_dot(i) = 0;
        }
    }

    VectorState y_dot;
    if(this->quaternion_mode){
        y_dot << this->constructPRotDotQ(y0, state.u0),
              uz_dot,
              state.uz,
              1,
              Eigen::Vector<DTYPE_t, STATE_NUM_R - STATE_NUM_Q>::Zero();
    }else{
        y_dot << this->constructPRotDotR(y0, state.u0),
              uz_dot,
              state.uz,
              1;
    }

    return y_dot;
}


VectorInput CosseratRod::evaluate(VectorInput x, bool use_fd_step){
    // If the solver jacobian is being built with a different step size
    if (use_fd_step) {
        this->integrator.setStepSize(this->fd_step);
    }
    else{
        this->integrator.setStepSize(this->h);
    }

    VectorInput psi_init = this->alphas + x.cwiseProduct(this->betas.cwiseAbs());

    if(this->quaternion_mode){
        this->y << constructPRotInitQ(psi_init(0)),
         x,
         psi_init,
         0,
         Eigen::Vector<DTYPE_t, STATE_NUM_R - STATE_NUM_Q>::Zero();
    }
    else{
        this->y << constructPRotInitR(psi_init(0)),
         x,
         psi_init,
         0;
    }

    this->J_eigen = -1;  // Reset maximum eigenvalue
    uint tube_end = TUBE_NUM - 1;  // Index to check for which tube end the integration is at

    int start_idx = this->idx;
    for(; this->idx < 2*TUBE_NUM; this->idx++){
        this->y = this->integrator.integrate(this->_derive, this->y, this->tube_lengths[this->idx], this->tube_lengths[this->idx + 1]);

        if(this->tube_lengths[this->idx + 1] == this->tube_ends[tube_end]){
            this->tube_states[tube_end] = this->y;
            tube_end--;
        }
    }
    this->idx = start_idx;

    if(this->calculate_J){
        this->J_eigen = std::max(this->J_eigen, this->calculateMaxJacobianEigen(this->y));
    }

    VectorInput u_forces;

    if(this->quaternion_mode){
        u_forces = this->y.segment(P_ROT_NUM_Q, TUBE_NUM);
    }
    else{
        u_forces = this->y.segment(P_ROT_NUM_R, TUBE_NUM);
    }

    return u_forces;
}


VectorInput CosseratRod::solve(VectorInput u_init_guess){
    this->func_counter = 0;
    this->calculate_J = false;

    auto start = std::chrono::high_resolution_clock::now();
    VectorInput res = this->solver.solve(this->_evaluate, u_init_guess);
    auto end = std::chrono::high_resolution_clock::now();

    auto runtime = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
    this->runtime = runtime.count();

    return res;
}
